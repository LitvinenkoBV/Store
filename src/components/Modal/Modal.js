import React from "react";
import "./Modal.scss";
import PropTypes from "prop-types";
import { useDispatch, useSelector } from "react-redux";
import {
  addProductsToBuyAction,
  removeProductsToBuyAction,
} from "../../store/products_to_buy/action";

const Modal = (props) => {
  const dispatch = useDispatch();
  const modal = useSelector((state) => state.modal);
  const arrProductsToBuy = useSelector((state) => state.productsToBuy);
  const { id, title, text, closeButton, closeModal } = props;

  const addProductsToBuy = () => {
    if (!arrProductsToBuy.includes(modal.currentProduct)) {
      localStorage.setItem(
        "arrProductsToBuy",
        JSON.stringify([...arrProductsToBuy, modal.currentProduct])
      );
      dispatch(
        addProductsToBuyAction([...arrProductsToBuy, modal.currentProduct])
      );
    } else {
      let arr = arrProductsToBuy.filter((product) => {
        if (JSON.stringify(product) !== JSON.stringify(modal.currentProduct)) {
          return product;
        }
      });
      localStorage.setItem("arrProductsToBuy", JSON.stringify(arr));
      dispatch(removeProductsToBuyAction(arr));
    }
    closeModal();
  };

  return (
    <div className="modal" onClick={() => props.closeModal()}>
      <div
        className="modal__content"
        onClick={(e) => e.stopPropagation()}
        key={id}
      >
        <div className="modal__title">{title}</div>
        {closeButton && (
          <button onClick={() => closeModal()} className="modal__close-button">
            +
          </button>
        )}
        <div className="modal__text">{text}</div>
        <div className="modal__button-confirm">
          <>
            {id === 0 && (
              <button
                onClick={() => addProductsToBuy()}
                className="modal__button-ok"
                value={true}
              >
                Так
              </button>
            )}
            {id === 1 && (
              <button
                onClick={() => addProductsToBuy()}
                className="modal__button-ok"
                value={true}
              >
                Так
              </button>
            )}

            <button
              onClick={() => closeModal()}
              className="modal__button-cancel"
              value={false}
            >
              Ні
            </button>
          </>
        </div>
      </div>
    </div>
  );
};

Modal.propTypes = {
  addProductsToBuy: PropTypes.func,
  closeButton: PropTypes.bool,
  closeModal: PropTypes.func,
  id: PropTypes.number,
  text: PropTypes.string,
  title: PropTypes.string,
};
Modal.defaultProps = {
  title: "Додати товар до кошику?",
};

export default Modal;
