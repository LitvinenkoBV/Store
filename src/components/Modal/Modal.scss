.modal {
  height: 100vh;
  width: 100vw;
  background-color: rgba(0, 0, 0, 0.4);
  position: fixed;
  top: 0;
  left: 0;
  display: flex;
  align-items: center;
  justify-content: center;

  &__content {
    position: relative;
    border-radius: 12px;
    background-color: white;
    width: 400px;
    background: #e74b3b;
  }
  &__title {
    background: #d34537;
    padding: 20px;
    width: 100%;
    color: white;
    font-size: 16px;
    border-radius: 12px 12px 0 0;
  }
  &__text {
    color: white;
    padding: 25px;
    line-height: 1.7;
    text-align: center;
  }
  &__close-button {
    transform: rotate(45deg);
    color: #3a3333;
    position: absolute;
    top: 5px;

    right: 15px;
    background: transparent;
    border: 0;
    font-size: 40px;
  }
  &__button-confirm {
    display: flex;
    justify-content: center;
  }
  &__button-ok,
  &__button-cancel {
    width: 80px;
    padding: 10px;
    background: #b3382b;
    color: white;
    border-radius: 3px;
    margin: 0px 10px 10px;
  }
}
