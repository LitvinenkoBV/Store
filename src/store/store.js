import { createStore, combineReducers, compose, applyMiddleware } from "redux";
import thunk from "redux-thunk";
import { productsReducer } from "./products/reducer";
import { getFavoriteProductsReducer } from "./favorite_products//reducer";
import { productsToBuyReducer } from "./products_to_buy/reducer";
import { modalReducer } from "./modal/reducer";

export const rootReducer = combineReducers({
  products: productsReducer,
  favoriteProducts: getFavoriteProductsReducer,
  productsToBuy: productsToBuyReducer,
  modal: modalReducer,
});

const devTools = window.__REDUX_DEVTOOLS_EXTENSION__
  ? window.__REDUX_DEVTOOLS_EXTENSION__()
  : (f) => f;

export const store = createStore(
  rootReducer,
  compose(applyMiddleware(thunk), devTools)
);
