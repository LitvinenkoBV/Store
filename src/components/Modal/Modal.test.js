import { render, screen } from "@testing-library/react";
import Modal from "./Modal";
import { Provider } from "react-redux";
import { legacy_createStore as createStore } from "redux";
import { rootReducer } from "../../store/store";

describe("<Modal/> component", () => {
  const store = createStore(rootReducer, {
    modal: { isOpen: false, currentProduct: {} },
    productsToBuy: [],
  });
  it("should render", () => {
    const { container } = render(
      <Provider store={store}>
        <Modal />
      </Provider>
    );
    expect(container.firstChild).toMatchSnapshot();
  });
  it("should render modal with title", () => {
    render(
      <Provider store={store}>
        <Modal title="Add to Card" />
      </Provider>
    );
    screen.getByText(/Add to Card/);
  });
});
