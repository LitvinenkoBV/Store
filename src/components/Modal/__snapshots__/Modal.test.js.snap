// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Modal/> component should render 1`] = `
<div
  class="modal"
>
  <div
    class="modal__content"
  >
    <div
      class="modal__title"
    >
      Додати товар до кошику?
    </div>
    <div
      class="modal__text"
    />
    <div
      class="modal__button-confirm"
    >
      <button
        class="modal__button-cancel"
        value="false"
      >
        Ні
      </button>
    </div>
  </div>
</div>
`;
